lib/collection/src/grouping/aggregator.rs

```diff
@@ -0,0 +1,425 @@
+}
+
+impl GroupsAggregator {
+    pub(super) fn new(groups: usize, group_size: usize, grouped_by: String) -> Self {
+        Self {
+            groups: HashMap::with_capacity(groups),
+            max_group_size: group_size,
+            grouped_by,
+            max_groups: groups,
+            full_groups: HashSet::with_capacity(groups),
+        }
+    }
+
+    /// Adds a point to the group that corresponds based on the group_by field, assumes that the point has the group_by field
+    fn add_point(&mut self, point: &ScoredPoint) -> Result<(), AggregatorError> {
+        if self.full_groups.len() == self.max_groups {
+            return Err(AllGroupsFull);
+        }
+
+        // if the key contains multiple values, grabs the first one
+        let group_key = point
+            .payload
+            .as_ref()
+            .and_then(|p| p.get_value(&self.grouped_by).into_iter().next().cloned())
+            .ok_or(KeyNotFound)
+            .and_then(GroupKey::try_from)?;
+
+        let group = self
+            .groups
+            .entry(group_key.clone())
+            .or_insert_with(|| HashSet::with_capacity(self.max_group_size));
+
+        group.insert(HashablePoint::minimal_from(point));
+
+        if group.len() == self.max_group_size {
+            self.full_groups.insert(group_key);
+        }
+
+        Ok(())
+    }
+
+    /// Adds multiple points to the group that they corresponds based on the group_by field, assumes that the points always have the grouped_by field, else it just ignores them
+    pub(super) fn add_points(&mut self, points: &[ScoredPoint]) {
+        for point in points {
+            if let Err(AllGroupsFull) = self.add_point(point) {
+                break;
+            }
+        }
+    }
+
+    #[cfg(test)]
+    pub(super) fn len(&self) -> usize {
+        self.groups.len()
+    }
+
+    // Gets the keys of the groups that have less than the max group size
+    pub(super) fn keys_of_unfilled_groups(&self) -> Vec<Value> {
+        self.groups
+            .iter()
+            .filter(|(_, hits)| hits.len() < self.max_group_size)
+            .map(|(key, _)| Value::from(key.clone()))
+            .collect()
+    }
+
+    // gets the keys of the groups that have reached the max group size
+    pub(super) fn keys_of_filled_groups(&self) -> Vec<Value> {
+        self.full_groups.iter().cloned().map(Value::from).collect()
+    }
+
+    /// Gets the ids of the already present points across all the groups
+    pub(super) fn ids(&self) -> HashSet<ExtendedPointId> {
+        self.groups
+            .iter()
+            .flat_map(|(_, hits)| hits.iter())
+            .map(|p| p.id())
+            .collect()
+    }
+
+    // Returns the best groups sorted by their best hit. The hits are sorted too
+    pub(super) fn distill(&self) -> Vec<Group> {
+        self.groups
+            .iter()
+            .map(|(k, v)| {
+                let hits = v
+                    .iter()
+                    .sorted()
```

Comment:
I think the `sorted` + reversing pattern could be optimized.
 
`sorted` requires to allocates the all collection

Might as well use `sorted_by` to sort it directly in descending order to skip the `reverse` iterator.

Reply:
Took me a little to get it. Both `sorted` and `sorted_by` consume the whole iterator, and both say
> The sorted iterator, if directly collected to a Vec, is converted without any extra copying or allocation cost.

I'm not sure if there is any difference here, but we can skip the `rev` usage if it helps, `rev` only traverses the iterator in reverse anyway.